xarray/coding/times.py

```diff
@@ -628,25 +628,37 @@ def _encode_datetime_with_cftime(dates, units: str, calendar: str) -> np.ndarray
     if cftime is None:
         raise ModuleNotFoundError("No module named 'cftime'")
 
+    dates = np.array(dates)
+
+    if dates.shape == ():
+        dates = dates.reshape(1)
+
     if np.issubdtype(dates.dtype, np.datetime64):
         # numpy's broken datetime conversion only works for us precision
         dates = dates.astype("M8[us]").astype(datetime)
 
-    def encode_datetime(d):
-        # Since netCDF files do not support storing float128 values, we ensure
-        # that float64 values are used by setting longdouble=False in num2date.
-        # This try except logic can be removed when xarray's minimum version of
-        # cftime is at least 1.6.2.
-        try:
-            return (
-                np.nan
-                if d is None
-                else cftime.date2num(d, units, calendar, longdouble=False)
-            )
-        except TypeError:
-            return np.nan if d is None else cftime.date2num(d, units, calendar)
+    # Find all the None position
+    none_position = np.equal(dates, None)
+
+    # Remove None from the dates and return new array
+    filtered_dates = dates[~none_position]
 
-    return np.array([encode_datetime(d) for d in dates.ravel()]).reshape(dates.shape)
+    # Since netCDF files do not support storing float128 values, we ensure
+    # that float64 values are used by setting longdouble=False in num2date.
+    # This try except logic can be removed when xarray's minimum version of
+    # cftime is at least 1.6.2.
+    try:
+        encoded_nums = cftime.date2num(
+            filtered_dates, units, calendar, longdouble=False
+        )
+    except TypeError:
+        encoded_nums = cftime.date2num(filtered_dates, units, calendar)
+
+    # Create a full matrix of NaN
+    # And fill the num dates in the not NaN or None position
+    result = np.full(dates.shape, np.nan)
```

Comment:
@antscloud This is a bit tricky. You are creating an array of floating point values here. Below `int64` is inserted for the breaking tests, effectively loosing precision.

A quick fix would be to initialize the array depending on dtype of `encoded_nums`. Something along the lines:

```suggestion
    nat = np.nan if encoded_nums.dtype.kind == "f" else np.iinfo(np.int64).min
    result = np.full(dates.shape, nat)
```

This at least fixes the tests. We would need to check, if the NaT-value (np.iinfo(np.int64).min) is good in all non floating point cases.

Reply:
Isn't this a better solution : 
```python
# Create a full matrix of NaN and dispatch the type based on the dtype of
# the encoded nums then fill the num dates in the not None position

# If no None is present, return the encoded nums type
# Otherwise, return the float if encoded_nums is float or object 
# if encoded_nums is different
if filtered_dates.size == none_position.size:
    _type = encoded_nums.dtype
else:
    _type = np.float64 if encoded_nums.dtype.kind == "f" else object

result = np.full(dates.shape, np.nan, dtype=_type)
result[np.nonzero(~none_position)] = encoded_nums
```
In my opinion it reproduces exactly the previous code and if you do this, we don't have to deal with 'pseudo NaN' value for integer and this avoid edge cases.
I tested it and all the tests seems to pass. What do you think ?

Reply:
Thanks @antscloud — unfortunately I think returning an object-dtype array is a non-starter for an encoding function.  

For now maybe we can immediately return `encoded_nums` in the case that `filtered_dates.size == none_position.size` (i.e. arrays that contain no missing values), and keep things essentially as in the code you initially committed for all other cases.  This would preserve the existing return types, and preserve the existing placeholder for missing values (i.e. `np.nan` instead of `np.iinfo(np.int64).min`).  Let me know if that makes sense.

Indeed, this does mean perpetuating possible precision loss in the case of arrays of dates with missing values, but we are not currently able to decode (or otherwise really do anything meaningful with) missing `cftime.datetime` values anyway: https://github.com/pydata/xarray/pull/7827#pullrequestreview-1608125736.  This is largely something that needs to be sorted out upstream: https://github.com/Unidata/cftime/issues/145.  

In another PR it may be worth deprecating (and eventually deleting) this old untested missing-value aspect of `cftime.datetime` encoding until that upstream issue is addressed.  Once that is done, we can try to handle this "the right way" and preserve roundtrip precision even in the presence of missing values.  @kmuehlbauer did a nice job of this for `datetime64[ns]` values in https://github.com/pydata/xarray/pull/7827.